## Modeling

In [63]:
import numpy as np
import pandas as pd 
data = pd.read_csv('Long_form.csv')
data.head(2)

Menu_Item  Westage  Count
0        Dosa    12.91    241
1  Idly, Vada    18.03    428

In [64]:
print(data.iloc[1,0])
type(data.iloc[1,0])

Idly, Vada


str

In [65]:
data.shape

(368, 3)

In [66]:
def convert(obj):
    lst = obj.split(', ')
    Final = []
    for dish in lst:
        Final.append(dish.replace(' ',''))

    return " ".join(Final)

In [67]:
data['Menu_Item'] = data['Menu_Item'].apply(convert)
data.sample(20)

Menu_Item  Westage  Count
158       Phulka MixVegMatar LaukiChanadal Gobi65 Rice    18.02    514
134  Poori DumAloo BeansCarrotPoriyal Rice Vathakol...    23.31    665
154    Phulka PeasMasala AlooMasalaWedges Rice Sambhar    19.31    554
84                                                Dosa    14.03    262
317  CholeBhature OnionMirchSalad Rice SnakeGourdKootu    13.33    508
42                                                Dosa    16.97    317
6                                       Pesarattu Upma    13.33    285
363                         LachaParatha VegKofta Rice    13.64    569
316                           Phulka VegCurry Rice Dal    12.65    513
312                                      SpecialDinner    19.13    880
262                                          SweetCorn     5.27    365
127  Phulka VegKurma BeetrootChannaPoriyal Rice Vat...    23.94    664
11                           MethiParatha ChannaMasala    17.57    340
102       Phulka MixVegMatar LaukiChanadal Gobi65 Rice    22.83    651
347                      Phulka PalakPaneer VegBiryani    12.91    501
56                                                Dosa    14.99    280
348                      Dosa Rice Sambhar AlooPodimas    15.82    609
57                                           Idly Vada    15.29    363
9                             ChowChowBath MysoreBonda    11.94    274
340                                      SpecialDinner    13.11    603

In [68]:
data.head(2)

Menu_Item  Westage  Count
0       Dosa    12.91    241
1  Idly Vada    18.03    428

In [69]:
type(data.iloc[1,0])

str

In [70]:
data.Menu_Item.value_counts()

Menu_Item
Dosa                                                                      13
SweetCorn                                                                 13
OnionPakoda                                                               13
Idly Vada                                                                 13
DahiVada                                                                  13
AlooSamosa                                                                12
Phulka ToorDal VegKurma Rice                                               7
Idly Podi Chutney Rice Sambhar                                             7
Phulka PaneerButterMasala Rice Sambhar                                     7
BananaBajji                                                                7
Maggi                                                                      7
Dosa Rice Sambhar AlooPodimas                                              7
MilletSnack                                                       

## Model 

In [71]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer()

In [72]:
vectors = cv.fit_transform(data['Menu_Item']).toarray()
vectors.shape

(368, 87)

## Splitting Training and Testing For Westage

In [73]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(vectors, data['Westage'], test_size=.15, random_state=len(data))

In [74]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(312, 87)
(56, 87)
(312,)
(56,)


## Selecting Different Models

### Linear Regression

In [ ]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

### Decision Tree Regressor

In [75]:
from sklearn.tree import DecisionTreeRegressor
model = DecisionTreeRegressor()

### Random Forest Regressor

In [ ]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor()

### Support Vector Regression

In [88]:
from sklearn.svm import SVR
model = SVR()

### Gradient Boosting Regressor
Gaussian Naive Byes

In [83]:
from sklearn.ensemble import GradientBoostingRegressor
model = GradientBoostingRegressor()

### Neural Network Regressor

In [93]:
from sklearn.neural_network import MLPRegressor
model = MLPRegressor()

In [94]:
model.fit(x_train, y_train)

e:\Setup\Python\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPRegressor()

In [95]:
y_train_pre = model.predict(x_train)
y_pre = model.predict(x_test)

# Evaluation

In [96]:
from sklearn.metrics import mean_squared_error, r2_score, confusion_matrix

#### Training Set

In [97]:
RMSE = mean_squared_error(y_train, y_train_pre)
R2 = r2_score(y_train, y_train_pre)
# Accuracy = model.score(x_train, y_train)


print(RMSE)
print(R2)
# print(Accuracy)

5.521357362319088
0.847915677292914


#### Testing Set

In [98]:
RMSE = mean_squared_error(y_test, y_pre)
R2 = r2_score(y_test, y_pre)
# Accuracy = model.score(x_test, y_test)


print(RMSE)
print(R2)
# print(Accuracy)

7.453949223457471
0.8452843879349432


# Results

The Results and accuracy scores of the models as summary

### Wastage

| Model | Train/Test | RMSE | R2 Score|
|-------|------------|------|---------|
| Linear Regression | Train | 5.387168726471041 | 0.8885819065769296 |
| Linear Regression | Test | 6.165576225981842 | 0.8598599395694546 |
| Decison Tree Regressor | Train | 3.822037457053275 | 0.8947229929393822 |
| Decison Tree Regressor | Test | 6.399925922950648 | 0.8671618994634029 |
| Random Forest Regressor | Train | 3.8300151552084847 | 0.8945032493616568 |
| Random Forest Regressor | Test | 6.279218965718302 | 0.8696673164187455 |
| Gradient Boosting Regressor | Train | 5.889639200602035 | 0.837771451830743 |
| Gradient Boosting Regressor | Train | 8.77539541398132 | 0.8178561951677401 |
| Support Vector Regressor | Train | 7.273659611899595 | 0.7996489770383194 |
| Support Vector Regressor | Test | 10.488283623342653 | 0.7823031561322215 |
| MPL Regressor | Train | 5.521357362319088 | 0.847915677292914 |
| MPL Regressor | Train | 7.453949223457471 | 0.8452843879349432 |

From the above Table we can say that Random Forest Regressor is the best Regression model for our Task of calculating wastage

### Daily Consumers
| Model | Train/Test | RMSE | R2 Score|
|-------|------------|------|---------|